NUX: Add Store segment to site-type step

niranjan-uma-shankar opened the pull request:
### Changes proposed in this Pull Request

* A new `Online Store` segment is added to the `site type` step
<img width="811" alt="screenshot 2018-11-19 at 7 21 30 pm" src="https://user-images.githubusercontent.com/1269602/48711165-7d535a00-ec30-11e8-8fa0-e8050f232ee9.png">

* The `Non-profit organization` segment is removed

* Selecting `Online Store` will redirect user to the new ecommerce flow


### Under the hood changes
 
* For the site type step, all constants (labels, descriptions et al) are stored as an object.

#### TODO

- [x] Merge #28680 before deploying this.

### Testing instructions

<!--
Add as many details as possible to help others reproduce the issue and test the fix.
"Before / After" screenshots can also be very helpful when the change is visual.
-->

#### Scenario 1

1. Enter the `Onboarding` flow AB test
2. In the Site Type step, select `Online Store` and click the Next button
3. Verfiy that you've entered the ecommerce flow ( the url changes to `/start/ecommerce/` )
     -  you should be taken to the Domain step after the site type step
     -  the About step is not shown
4. Hit the Back button, and verify that you're taken to the Site Type step(`/start/ecommerce/site-type`)
     - you should not be taken to the About step
5. Change site type to something else like Blog, and then click the Next button
6. Verify that you're no longer in the ecommerce flow ( the url does not have `ecommerce` in it).
7. You should now be in the Site Topic step. Proceed Next.
8. You should now be in the About step. Proceed Next.
9. You should now be in the Domains step.
10. Keep pressing the Back button till you reach the Site Type step. Verify that all the steps you traversed through appear in expected order.


#### Scenario 2
1. In the `Main` flow, go through the signup flow process
2. Go through both the ecommerce flow (by selecting sell option in site goals checkbox in About step) and the main flow.
3. Verfiy that changes introduced by this PR do no affect the main flow.

#### Scenario 3
1. Enter the signup flow with a query parameter matching one of the site type slugs, e.g: `http://calypso.localhost:3000/start/user?site_type=online-store`. 
2. Verify that the corresponding site type radio option is selected when the segment page loads.

#### Scenario 4
Verify that the checklist that loads in Calypso matches the segment. E.g. `business` and `blog` site types have different checklists.

_If you have tried more tests, please add to the list_

niranjan-uma-shankar commented:
![](https://github.trello.services/images/mini-trello-icon.png) [Signup: Add 'store' segment to `site-type` step](https://trello.com/c/NWxjqGV2/257-signup-add-store-segment-to-site-type-step)

fditrapani commented:
Thanks for adding this. 

> Verfiy that you've entered the ecommerce flow ( the url changes to /start/ecommerce/about/ )

the URL changes for me but it's not showing the ecommerce flow I see if I go through our main flow and only select the "Sell products or collect payments" Goal. When I do that in the main flow I am presented with a domain search as my second step vs the about field.

niranjan-uma-shankar commented:
@fditrapani Thanks for the review. I've fixed it in [9a0933c](https://github.com/Automattic/wp-calypso/pull/28672/commits/9a0933c8138d866c2d17a5acc3f1a527a1779025).

fditrapani commented:
Cool Thanks @niranjan-uma-shankar. The steps now work the same as the ecommerce flow on main. With this update, I spotted another bug. If you press back on the domain step it takes you back to the main about versus the segment step in the onboarding flow.

michaeldcain commented on client/signup/utils.js:
```diff
@@ -146,19 +146,19 @@ export function getThemeForDesignType( designType ) {
 
 export function getThemeForSiteType( siteType ) {
-	if ( 'business' === dasherize( siteType ) ) {
+	if ( siteType && siteType === get( allSiteTypes, 'business.value', '' ) ) {
```
We can probably leave these hardcoded to their siteType IDs for the time being (i.e. `'business' === siteType`).

michaeldcain commented on client/signup/steps/site-type/index.jsx:
```diff
@@ -109,6 +118,17 @@ export default connect(
 			dispatch( setSiteType( siteTypeValue ) );
+
+			let nextFlowName = ownProps.flowName;
+			if ( siteTypeValue === get( allSiteTypes, 'store.value', '' ) ) {
```
We can probably leave these hardcoded to their siteType IDs for the time being (i.e. 'online-store' === siteType).

michaeldcain commented on client/signup/steps/site-type/index.jsx:
```diff
@@ -26,45 +27,53 @@ import FormRadio from 'components/forms/form-radio';
-		const radioOptions = allSiteTypes.map( elem => (
-			<FormLabel className="site-type__option" key={ elem.type }>
+	renderRadioOptions() {
+		return Object.values( allSiteTypes ).map( elem => (
```
If we go with the `array/object` structure proposed above, this could just be `allSiteTypes.map( elem => ( //...`

niranjan-uma-shankar commented on client/lib/signup/site-type.js:
```diff
@@ -13,48 +14,42 @@ import i18n from 'i18n-calypso';
-export const allSiteTypes = [
-	{
-		type: i18n.translate( 'Blog' ),
+export const allSiteTypes = {
```
Thanks for the inputs @michaeldcain.

> Also, to make the object simpler, I think we can use an 'id' value as both the 'value' and 'queryParam'. I don't think there's a need to have two versions (the radio form value is hidden from the user), and I think uncapitalized and statically dashed is the way to go.

Combining both fields is a good idea. I splitted it into two fields because I thought maybe we should store the unmodified site type value (i.e undashed, if that is a word) in the backend. However, on thinking through this again, I see no issues particularly in storing the uncapitalized and dashed value as the user's selection. So, the `id` field will serve the dual purpose of being the value of the site type selection, and also the slug that is used in the vertical landing pages e.g `?vertical=bar&site_type=<id>`.

> If we go the route of having our "slug" as a value inside the object (in this current case, value: 'Blog'), then I think we can keep this as an array of objects

The main reason why I went with the approach of objects ibstead of array is that the object key will remain unchanged, whereas `label` and `slug` values could change over time. So, down the line if we decide to change `Online Store` to say `Ecommerce Store`, and the corresponding slug or `id` field changes to `ecommerce-store`, the references to this site type in the code can still remain as `allSiteTypes.store`.

We're using specific references in a few places. E.g:
- In `getThemeForSiteType()`, we assign a theme depending on the site type value. 
```
if ( siteType && siteType === get( allSiteTypes, 'education.value', '' ) ) {
    return 'pub/twentyfifteen';
}
```

^^^ If `allSiteTypes` is made an array, could we write this without having to hardcode the site type value?

- In a few places([#](https://github.com/Automattic/wp-calypso/pull/28182/files#diff-c1e6c4d31974e650b129e726b3af771aR175), [#](https://github.com/Automattic/wp-calypso/blob/master/client/signup/steps/site-topic/index.jsx#L113), [#](https://github.com/Automattic/wp-calypso/pull/28604/files#diff-7949db3a74b97e79a5ff08e5b794a6d8R570)), the copy depends on the site type value, and we're hardcoding the site type value in those places. The object method helps rid us of the hardcoding and touching multiple files for segment copy changes, especially given that the segments could probably face changes in the coming weeks (we may add some, remove or edit some?).

In the above two cases, is it possible to compare the site type value without hardcoding if we use the array method? 

It's possible that I'm overthinking this, so do point me in the right direction if so 😄.

adamziel replied:
> So, down the line if we decide to change Online Store to say Ecommerce Store, and the corresponding slug or id field changes to ecommerce-store, the references to this site type in the code can still remain as allSiteTypes.store.

I'd say in that case we would only change the name and keep ID the same, even if that would mean it says something a bit different

michaeldcain replied:
@niranjan-uma-shankar: I like how you're thinking on the future code stability. I'm fine with using the object key approoach, but can we add `theme` and `designType` params to the objects and the change `getThemeForSiteType` and `getDesignTypeForSiteType` to helper functions (inside `/lib/signup/site-type.js`) that just get the values.

The following feels more like something that would be returned from the upcoming `/segments/` endpoint.

```
allSiteTypes = {
  'blog': {
    id: 'blog',
    label: i18n.translate( 'Blog' ),
    description: i18n.translate( 'Share and discuss ideas, updates, or creations.' ),
    theme: 'pub/radcliffe-2',
    designType: 'posts',
  },
  //...
};
```

niranjan-uma-shankar replied:
> but can we add theme and designType params to the objects and the change getThemeForSiteType and getDesignTypeForSiteType to helper functions (inside /lib/signup/site-type.js) that just get the values.

Yes indeed. We can certainly do that, was thinking the same. 👍

niranjan-uma-shankar replied:
After another round of discussions, we've decided to make `allSiteTypes` an array of objects as it best addresses all our needs. So, the format will  be:
```
allSiteTypes = [
  {
    id: 'blog',     
    slug: 'blog',  //used in vertical landing pages
    label: i18n.translate( 'Blog' ),
    description: i18n.translate( 'Share and discuss ideas, updates, or creations.' ),
    theme: 'pub/radcliffe-2',
    designType: 'posts',
  },
  //...
];
```

This ensures that iterating over allSiteTypes using `Array.prototype.map()` will always be [in order](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Array/map#Description).

niranjan-uma-shankar replied:
This is fixed in [da96beb](https://github.com/Automattic/wp-calypso/pull/28672/commits/da96beb444e5fe822efc68f8d0047d7c17742515)

michaeldcain commented:
@niranjan-uma-shankar: can you update these keys with the new helpers, too?

https://github.com/Automattic/wp-calypso/pull/28475#discussion_r234844579

michaeldcain commented on client/signup/steps/site-type/index.jsx:
```diff
@@ -109,6 +118,17 @@ export default connect(
 	( dispatch, ownProps ) => ( {
 		submitStep: ( siteTypeValue, themeRepo ) => {
 			dispatch( setSiteType( siteTypeValue ) );
```
@niranjan-uma-shankar: can we add an explicit `calypso_signup_actions_submit_site_type` tracks event here, like [we ended up doing for siteTopic](https://github.com/Automattic/wp-calypso/blob/master/client/signup/steps/site-topic/index.jsx#L156)? In our funnel audit, I think it would make things clearer.

niranjan-uma-shankar replied:
Fixed in [bf39dc3](https://github.com/Automattic/wp-calypso/pull/28672/commits/bf39dc38849da0a1424461f60a5286740827f9bd)

niranjan-uma-shankar commented:
I've changed `allSiteTypes` to an Array. Also, I've fixed the Back button @fditrapani . I've expanded the Testing Instructions to include a few more steps for the Back button testing. 

Bringing to your notice a behaviour. When you enter the Onboarding flow, the total number of steps will be shown as `6`:

<img width="403" alt="screenshot 2018-11-21 at 11 41 26 pm" src="https://user-images.githubusercontent.com/1269602/48860694-9e11e000-ede7-11e8-88f2-156f48325f43.png">

The total steps count does not change if you proceed with any segment other than the Store segment. 
If you select `Online store`, the total step count changes to `4`:

<img width="587" alt="screenshot 2018-11-21 at 11 41 37 pm" src="https://user-images.githubusercontent.com/1269602/48860769-d9141380-ede7-11e8-94a6-6f2364fb2cfb.png">

The total steps count is correct in all cases but do you think it's a UX issue that they suddenly change from `6` to `4`? I don't know if there's any way to change this behaviour(unless we add more steps to the ecommerce flow), but just bringing to your notice.

niranjan-uma-shankar commented:
![](https://github.trello.services/images/mini-trello-icon.png) [Site type: Explore if site type array can be converted to an object](https://trello.com/c/OOoNiwjV/245-site-type-explore-if-site-type-array-can-be-converted-to-an-object)

fditrapani commented:
I just checked out a copy of the branch locally, and tried this on calypso.live, and found the back button wasn't showing on the second step. I also noticed the steps indicated there were only two steps:

![image](https://user-images.githubusercontent.com/6981253/48864419-78aab500-ed9a-11e8-800b-5d271516c39f.png)

Something like this happened before, am I getting a cached version?

niranjan-uma-shankar commented:
> I just checked out a copy of the branch locally, and tried this on calypso.live, and found the back button wasn't showing on the second step. I also noticed the steps indicated there were only two steps:
>Something like this happened before, am I getting a cached version?

This is a valid observation. Sorry about this, I was testing the signup flow from the logged out interface and missed to test the Add Site option in the logged in interface. 

I've fixed this in [293468a](https://github.com/Automattic/wp-calypso/pull/28672/commits/293468a987905f1060732deb37b93a83ae22d0c3).

niranjan-uma-shankar commented:
There's a bug I noticed in `Add New Site` in production, that affects the Onboarding flow too. This bug is seen in the master branch.

In the logged in interface, if I perform the following step:
(1) While logged in, go to `Add a New Site` - _/jetpack/new?ref=calypso-selector_
(2) Go through the ecommerce flow 
(3) Drop off at the Plans step - navigate away to a different part of calypso or close tab but continue to remain logged in
(4) Go to `Add a New Site` again - _/jetpack/new?ref=calypso-selector/_ - and click _Start Now_ button

It directly takes to the Plans step but check the URL - it's not in the ecommerce flow any longer. I would expect that since I dropped off the Plans step in the ecommerce flow, I should still be in the same flow when I'm taken there.

A short screen grab of the issue - https://cloudup.com/cC_f82jHuA2

This issue manifests here too - if we select Online Store in the segment, and then drop off at the Plans step in the Add a New Site flow, then navigating back to Add New Site no longer puts us in the ecommerce flow. I need to investigae the cause, but maybe this could be taken as a separate PR? 
cc @adamziel @fditrapani @michaeldcain

michaeldcain commented:
> I need to investigae the cause, but maybe this could be taken as a separate PR?

@niranjan-uma-shankar: great catch - can you open a bug report for it? We can track it in a separate card.

michaeldcain commented:
> The total steps count is correct in all cases but do you think it's a UX issue that they suddenly change from 6 to 4? I don't know if there's any way to change this behaviour(unless we add more steps to the ecommerce flow), but just bringing to your notice.

@niranjan-uma-shankar: I think this is fine behavior as long as the step count accurately reflects the number of steps the flow has at that point.

michaeldcain commented:
@niranjan-uma-shankar: can you check the failing test?

niranjan-uma-shankar commented:
> @niranjan-uma-shankar: can you check the failing test?

@michaeldcain I've pushed in a partial fix for this in [90f9550](https://github.com/Automattic/wp-calypso/pull/28672/commits/90f955011defbfca1c1b4da4cd881ae6dbf600aa). 

`client/state/signup/progress/test/reducer.js` and `client/lib/signup/test/progress-store.js` are failing, the fix for which I've pushed in two separate PRs since they touch important files. 

Merging #28783  and #28784 should fix the failing tests on this PR.

niranjan-uma-shankar commented:
> @niranjan-uma-shankar: great catch - can you open a bug report for it? We can track it in a separate card.

Bug reported in #28785

adamziel commented on client/signup/config/flows-pure.js:
```diff
@@ -241,12 +241,22 @@ export function generateFlows( { getSiteDestination = noop, getPostsDestination
+		// please copy the same changes to ecommerce-onboarding flow too
 		flows.ecommerce = {
-			steps: [ 'about', 'domains', 'plans', 'user' ],
+			steps: [ 'user', 'about', 'domains', 'plans' ],
```
yup!

adamziel commented on client/signup/steps/about/index.jsx:
```diff
@@ -213,10 +209,12 @@ class AboutStep extends Component {
-			designType = getDesignTypeForSiteType( siteType, flowName );
+			themeRepo =
+				getSiteTypePropertyValue( 'slug', siteType, 'theme' ) || 'pub/independent-publisher-2';
+			designType =
```
I'm a bit confused by this bit: `a ? b : c || d` is not clear, what's the operator precedence there? Can it be more explicit? It's okay to use `if` even if it's longer

niranjan-uma-shankar replied:
Thanks, fixed in [8ce58b6](https://github.com/Automattic/wp-calypso/pull/28672/commits/8ce58b6eea60717f8758c44edef851aaa3e13e38.)

adamziel commented on client/state/signup/progress/reducer.js:
```diff
@@ -51,8 +53,23 @@ function processStep( state, { step } ) {
+	flowName =
+		'onboarding' === abtest( 'improvedOnboarding' ) && 'ecommerce' === flowName
+			? flowName + '-onboarding'
+			: flowName;
```
That's one long ternary expression! Can we make it something like `if(x) { flowName = 'ecommerce-onboarding'; }`?

niranjan-uma-shankar replied:
Thanks, fixed in [8ce58b6](https://github.com/Automattic/wp-calypso/pull/28672/commits/8ce58b6eea60717f8758c44edef851aaa3e13e38.)

adamziel commented on client/signup/steps/site-type/index.jsx:
```diff
@@ -115,6 +120,22 @@ export default connect(
+				nextFlowName =
+					'ecommerce' === ownProps.flowName && ownProps.previousFlowName
+						? ownProps.previousFlowName
+						: ownProps.flowName;
```
Can we reduce this to just `} else if ('ecommerce' === ownProps.flowName && ownProps.previousFlowName) { nextFlowName = ownProps. previousFlowName; }`?

niranjan-uma-shankar replied:
Thanks, fixed in [8ce58b6](https://github.com/Automattic/wp-calypso/pull/28672/commits/8ce58b6eea60717f8758c44edef851aaa3e13e38.)

adamziel commented on client/signup/config/flows.js:
```diff
@@ -226,6 +226,14 @@ const Flows = {
+			flow = Flows.removeStepFromFlow( 'about', flow );
+			return Flows.insertStepIntoFlow( 'site-type', flow, afterStep );
+		}
+
```
Do you think it would make sense to redirect users to the `ecommerce-onboarding` flow instead of dynamically updating the `ecommerce` flow? One downside is that it would change the URL to `/start/ecommerce-onboarding/domains`

niranjan-uma-shankar replied:
A few reasons why I chose to not redirect the user to a new ecommerce flow:
1) As you rightly said,  the URL will change to `ecommerce-onboarding` flow. 
2) The flow name `ecommerce` is hardcoded in a few places ([#](https://github.com/Automattic/wp-calypso/blob/affe06d124b57f4aac9e2d4e935f200692cd4c78/client/signup/steps/plans/index.jsx#L62), [#](https://github.com/Automattic/wp-calypso/blob/9ede582d81a4641524328ebe4bbd27be37c020ff/client/signup/utils.js#L210), [#](https://github.com/Automattic/wp-calypso/blob/2bc79d736f3bdd6a07374b8ed66c33af72002335/client/signup/steps/domains/index.jsx#L358), [#](https://github.com/Automattic/wp-calypso/blob/c46187dfab764fbe1cc122d8a30dc9faded5b162/client/signup/steps/about/index.jsx#L243)). So, adding a new flow will require changes across more files. The present fix touches the least number of places.
3) Future proofing: if we redirect users to the `ecommerce-onboarding` flow, then we'll need to maintain two flows - `ecommerce` and `ecommerce-onboarding` - in all future additions to the codebase that refer to flowName. I think this makes it more prone to errors. In the present scenario, it's future proofed - all code references to flowName can remain only to `ecommerce`.

Are there any downsides to the approach of dynamically updating the `ecommerce` flow that I haven't considered? Or reasons why it's less preferred?

adamziel commented:
@niranjan-uma-shankar I added a few comments, mostly nitpicks but there's one higher-level question too

fditrapani commented:
I tested this out on Calypso live and it looks good.

taggon commented on client/lib/signup/site-type.js:
```diff
@@ -15,46 +16,55 @@ import i18n from 'i18n-calypso';
-		.replace( / /g, '-' )
-		.replace( /-+/, '-' );
-}
+	const siteTypeProperties = find( allSiteTypes, object => {
```
This line can be simplified to `find( allSiteTypes, { [key]: value } );`

niranjan-uma-shankar replied:
Good suggestion, thanks!

niranjan-uma-shankar replied:
@taggon  Implemented in [283c9e0](https://github.com/Automattic/wp-calypso/pull/28672/commits/283c9e0834d57dcf197d15e93f907e3e6d0b0fae)
